yarn/src/history/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,1328 @@
+import scala.collection.mutable
+
+import com.codahale.metrics.{Counter, MetricRegistry, Timer}
+import org.apache.hadoop.security.UserGroupInformation
+import org.apache.hadoop.yarn.api.records.timeline.{TimelineDomain, TimelineEntity, TimelineEvent}
+import org.apache.hadoop.yarn.api.records.{ApplicationAttemptId, ApplicationId}
+import org.apache.hadoop.yarn.client.api.TimelineClient
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+
+import org.apache.spark.deploy.history.yarn.YarnTimelineUtils._
+import org.apache.spark.metrics.source.Source
+import org.apache.spark.scheduler._
+import org.apache.spark.scheduler.cluster.{SchedulerExtensionService, SchedulerExtensionServiceBinding}
+import org.apache.spark.util.{SystemClock, Utils}
+import org.apache.spark.{Logging, SparkContext}
+
+/**
+ * A Yarn Extension Service to post lifecycle events to a registered YARN Timeline Server.
+ *
+ * Posting algorithm
+ *
+ * 1. The service subscribes to all events coming from the Spark Context.
+ * 1. These events are serialized into JSON objects for publishing to the timeline service through
+ * HTTP(S) posts.
+ * 1. Events are buffered into `pendingEvents` until a batch is aggregated into a
+ * [[TimelineEntity]] for posting.
+ * 1. That aggregation happens when a lifecycle event (application start/stop) takes place,
+ * or the number of pending events in a running application exceeds the limit set in
+ * `spark.hadoop.yarn.timeline.batch.size`.
+ * 1. Posting operations take place in a separate thread from the spark event listener.
+ * 1. If an attempt to post to the timeline server fails, the service sleeps and then
+ * it is re-attempted after the retry period defined by
+ * `spark.hadoop.yarn.timeline.post.retry.interval`.
+ * 1. If the number of events buffered in the history service exceed the limit set in
+ * `spark.hadoop.yarn.timeline.post.limit`, then further events other than application start/stop
+ * are dropped.
+ * 1. When the service is stopped, it will make a best-effort attempt to post all queued events.
+ * the call of [[stop()]] can block up to the duration of
+ * `spark.hadoop.yarn.timeline.shutdown.waittime` for this to take place.
+ * 1. No events are posted until the service receives a [[SparkListenerApplicationStart]] event.
+ *
+ * If the spark context has a metrics registry, then the internal counters of queued entities,
+ * post failures and successes, and the performance of the posting operation are all registered
+ * as metrics.
+ *
+ * The shutdown logic is somewhat convoluted, as the posting thread may be blocked on HTTP IO
+ * when the shutdown process begins. In this situation, the thread continues to be blocked, and
+ * will be interrupted once the wait time has expired. All time consumed during the ongoing
+ * operation will be counted as part of the shutdown time period.
+ */
+private[spark] class YarnHistoryService extends SchedulerExtensionService
+  with Logging with Source {
```

Comment:
Can we make `Source` related things into a sub-class or separated class. Here in this class there's so many class parameters, it is not easy to understand and track the state.